File: api/src/open_hallucination_index/adapters/outbound/knowledge_sources/crossref.py
"""
Crossref API Adapter
====================

Queries Crossref for DOI metadata and scholarly works.
https://www.crossref.org/documentation/retrieve-metadata/rest-api/
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any
from urllib.parse import quote

from open_hallucination_index.adapters.outbound.knowledge_sources.base import (
    HTTPKnowledgeSource,
)
from open_hallucination_index.domain.entities import Evidence, EvidenceSource

if TYPE_CHECKING:
    from open_hallucination_index.domain.entities import Claim

logger = logging.getLogger(__name__)


class CrossrefAdapter(HTTPKnowledgeSource):
    """
    Adapter for Crossref REST API.

    Crossref is a DOI registration agency providing metadata
    for scholarly publications.
    """

    def __init__(
        self,
        base_url: str = "https://api.crossref.org",
        email: str | None = None,
        timeout: float = 30.0,
    ) -> None:
        super().__init__(base_url=base_url, timeout=timeout)
        self._email = email  # For polite pool

    @property
    def source_name(self) -> str:
        return "Crossref"

    @property
    def evidence_source(self) -> EvidenceSource:
        return EvidenceSource.CROSSREF

    async def connect(self) -> None:
        """Initialize with custom headers."""
        await super().connect()
        if self._client and self._email:
            self._client.headers["mailto"] = self._email

    async def health_check(self) -> bool:
        """Check Crossref API health."""
        if not self._client:
            return False
        try:
            response = await self._client.get("/works", params={"rows": 1})
            return response.status_code == 200
        except Exception:
            return False

    async def find_evidence(self, claim: Claim) -> list[Evidence]:
        """Find scholarly evidence for a claim."""
        if not self._available:
            return []

        evidences: list[Evidence] = []
        search_term = claim.subject or claim.text[:100]

        try:
            works = await self._search_works(search_term, limit=5)

            for work in works:
                title = self._get_title(work)
                if not title:
                    continue

                # Build content
                authors = self._format_authors(work.get("author", []))
                content = f"{title}\n\nAuthors: {authors}"

                abstract = work.get("abstract", "")
                if abstract:
                    # Strip JATS XML tags if present
                    import re

                    abstract = re.sub(r"<[^>]+>", "", abstract)
                    content += f"\n\nAbstract: {abstract[:800]}"

                published = self._get_publication_date(work)
                if published:
                    content += f"\n\nPublished: {published}"

                container = work.get("container-title", [])
                if container:
                    content += f"\nJournal: {container[0]}"

                doi = work.get("DOI", "")

                evidences.append(
                    self._create_evidence(
                        content=content,
                        source_id=f"crossref:{doi}",
                        source_uri=f"https://doi.org/{doi}" if doi else "",
                        similarity_score=0.85,
                        structured_data={
                            "title": title,
                            "doi": doi,
                            "type": work.get("type", ""),
                            "publisher": work.get("publisher", ""),
                            "is_referenced_by_count": work.get("is-referenced-by-count", 0),
                        },
                    )
                )

            logger.debug(f"Found {len(evidences)} Crossref evidences")
            return evidences

        except Exception as e:
            logger.warning(f"Crossref search failed: {e}")
            return []

    async def search(self, query: str, limit: int = 5) -> list[dict[str, Any]]:
        """Search Crossref works."""
        if not self._available:
            return []
        return await self._search_works(query, limit)

    async def _search_works(self, query: str, limit: int = 5) -> list[dict[str, Any]]:
        """Search for works via Crossref."""
        try:
            response = await self._client.get(
                "/works",
                params={
                    "query": query,
                    "rows": limit,
                    "sort": "relevance",
                },
            )
            response.raise_for_status()
            data = response.json()
            return data.get("message", {}).get("items", [])
        except Exception as e:
            logger.warning(f"Crossref search error: {e}")
            return []

    async def get_work_by_doi(self, doi: str) -> dict[str, Any] | None:
        """Get work metadata by DOI."""
        try:
            encoded_doi = quote(doi, safe="/")
            response = await self._client.get(f"/works/{encoded_doi}")
            if response.status_code == 200:
                data = response.json()
                return data.get("message", {})
            return None
        except Exception:
            return None

    def _get_title(self, work: dict[str, Any]) -> str:
        """Extract title from work."""
        titles = work.get("title", [])
        return titles[0] if titles else ""

    def _format_authors(self, authors: list[dict]) -> str:
        """Format author list."""
        names = []
        for author in authors[:5]:
            given = author.get("given", "")
            family = author.get("family", "")
            if family:
                name = f"{given} {family}".strip() if given else family
                names.append(name)

        if len(authors) > 5:
            names.append(f"and {len(authors) - 5} more")

        return ", ".join(names) if names else "Unknown"

    def _get_publication_date(self, work: dict[str, Any]) -> str:
        """Extract publication date."""
        published = work.get("published", {}) or work.get("created", {})
        date_parts = published.get("date-parts", [[]])
        if date_parts and date_parts[0]:
            parts = date_parts[0]
            if len(parts) >= 3:
                return f"{parts[0]}-{parts[1]:02d}-{parts[2]:02d}"
            elif len(parts) >= 2:
                return f"{parts[0]}-{parts[1]:02d}"
            elif len(parts) >= 1:
                return str(parts[0])
        return ""

File: src/ohi-mcp-server/src/sources/crossref.ts
/**
 * Crossref Source
 * ===============
 * 
 * DOI metadata and scholarly publications.
 */

import { BaseSource, SearchResult } from "./base.js";
import { httpClient } from "../utils/http-client.js";

interface CrossrefWork {
  DOI?: string;
  title?: string[];
  author?: Array<{ given?: string; family?: string }>;
  abstract?: string;
  published?: { "date-parts"?: number[][] };
  "container-title"?: string[];
  publisher?: string;
  type?: string;
  "is-referenced-by-count"?: number;
}

interface CrossrefResponse {
  message?: {
    items?: CrossrefWork[];
  };
}

export class CrossrefSource extends BaseSource {
  name = "crossref";
  description = "Crossref DOI metadata";
  private email?: string;

  constructor(email?: string) {
    super("https://api.crossref.org");
    this.email = email;
  }

  /**
   * Remove HTML-like tags from an abstract string in a robust way by
   * repeatedly stripping `<...>` patterns until the string stabilizes.
   */
  private sanitizeAbstract(raw: string): string {
    let previous: string;
    let current = raw;
    do {
      previous = current;
      // Recreate the global regex each iteration to avoid stateful `lastIndex`.
      const tagRegex = /<[^>]+>/g;
      current = current.replace(tagRegex, "");
    } while (current !== previous);
    return current;
  }

  async healthCheck(): Promise<boolean> {
    try {
      const response = await httpClient.get(`${this.baseUrl}/works`, {
        params: { rows: 1 },
      });
      return response.status === 200;
    } catch {
      return false;
    }
  }

  async search(query: string, limit = 5): Promise<SearchResult[]> {
    const doiMatch = query.match(/10\.\d{4,9}\/[^\s]+/i);
    if (doiMatch?.[0]) {
      const byDoi = await this.getByDOI(doiMatch[0]);
      if (byDoi) return [byDoi];
    }

    const compacted = this.compactQuery(query) || query;
    // Crossref polite pool: add mailto parameter for higher rate limits (50 req/sec vs 1 req/sec)
    // https://github.com/CrossRef/rest-api-doc#good-manners--more-reliable-service
    const params: Record<string, string | number> = {
      query: compacted,
      rows: limit,
      sort: "relevance",
    };
    
    if (this.email) {
      params.mailto = this.email;
    }

    const response = await httpClient.get<CrossrefResponse>(`${this.baseUrl}/works`, {
      params,
    });

    return (response.data.message?.items || []).map((work) => {
      const title = work.title?.[0] || "Untitled";
      const authors = (work.author || [])
        .slice(0, 3)
        .map((a) => `${a.given || ""} ${a.family || ""}`.trim())
        .join(", ");

      const abstract = work.abstract
        ? this.sanitizeAbstract(work.abstract).slice(0, 800)
        : "";

      const pubDate = this.formatDate(work.published?.["date-parts"]?.[0]);

      return {
        source: this.name,
        title,
        content: `Authors: ${authors || "Unknown"}\n${pubDate ? `Published: ${pubDate}\n` : ""}${work["container-title"]?.[0] ? `Journal: ${work["container-title"][0]}\n` : ""}\n${abstract}`,
        url: work.DOI ? `https://doi.org/${work.DOI}` : undefined,
        metadata: {
          doi: work.DOI,
          type: work.type,
          publisher: work.publisher,
          citation_count: work["is-referenced-by-count"],
        },
        score: work["is-referenced-by-count"],
      };
    });
  }

  async getByDOI(doi: string): Promise<SearchResult | null> {
    try {
      // Add polite pool mailto parameter if email is configured
      const params: Record<string, string> = {};
      if (this.email) {
        params.mailto = this.email;
      }
      
      const response = await httpClient.get<{ message?: CrossrefWork }>(
        `${this.baseUrl}/works/${encodeURIComponent(doi)}`,
        { params }
      );
      
      const work = response.data.message;
      if (!work) return null;

      const title = work.title?.[0] || doi;
      const authors = (work.author || [])
        .map((a) => `${a.given || ""} ${a.family || ""}`.trim())
        .join(", ");

      return {
        source: this.name,
        title,
        content: `DOI: ${doi}\nAuthors: ${authors}\nPublisher: ${work.publisher || "Unknown"}`,
        url: `https://doi.org/${doi}`,
        metadata: {
          doi,
          type: work.type,
          citation_count: work["is-referenced-by-count"],
        },
      };
    } catch {
      return null;
    }
  }

  private formatDate(parts?: number[]): string {
    if (!parts || parts.length === 0) return "";
    if (parts.length >= 3) return `${parts[0]}-${String(parts[1]).padStart(2, "0")}-${String(parts[2]).padStart(2, "0")}`;
    if (parts.length >= 2) return `${parts[0]}-${String(parts[1]).padStart(2, "0")}`;
    return String(parts[0]);
  }
}
